test/provision_test.py

```diff
@@ -0,0 +1,108 @@
+import pytest
+import json
+
+from autoprotocol.protocol import Protocol
+
+from test.test_util import TestUtils
+
+
+class TestProvision(object):
+    p = Protocol()
+    w1 = p.ref("w1", None, cont_type="96-pcr", discard=True) \
+        .well(0).set_volume("2:microliter")
+
+    def test_provision_well_capacity(self):
+        self.p.provision("rs17gmh5wafm5p", self.w1, "50:microliter")
+        actual_protocol_as_json = json.dumps(self.p.as_dict(), indent=2, sort_keys=True)
+        expected_protocol_as_json = TestUtils.read_json_file('expected_provision.json')
+
+        assert (expected_protocol_as_json == actual_protocol_as_json)
+
+    def test_attempt_to_provision_more_than_well_capacity(self):
+        with pytest.raises(ValueError):
+            self.p.provision("rs17gmh5wafm5p", self.w1, "500:microliter")
+
+    # TODO - If all container-types defined in `container_type.py` are of type Well, then code lines
+    # TODO - between 4960-4962 in protocol.py doesn't make any sense
+    # def test_attempt_to_provision_with_unsupported_container_type(self):
+    #     p = Protocol()
+    #     w1 = p.ref("w1", None, cont_type="micro-1.5", discard=True) \
+    #         .well(0).set_volume("2:microliter")
+    #     with pytest.raises(TypeError):
+    #         p.provision("rs17gmh5wafm5p", w1, "50:microliter")
+
+    def test_with_invalid_resourceId(self):
+        with pytest.raises(TypeError):
+            self.p.provision(100, self.w1, "50:microliter")
+
+    def test_with_destinations_count_not_same_as_volumes(self):
+        volumes = ["50:microliter", "20:microliter"]
+        with pytest.raises(RuntimeError):
+            self.p.provision("rs17gmh5wafm5p", self.w1, volumes)
+
+    # TODO - This test should pass as there is no definition of `ml` in http://autoprotocol.org/specification/.
+    # TODO - Since the current code is able to translate `ml` to `microliter`, this test case is failing
+    # def test_with_invalid_unit(self):
+    #     with pytest.raises(TypeError):
+    #         self.p.provision("rs17gmh5wafm5p", self.w1, "20:ml")
```

Comment:
ah. this is expected behavior currently. The Units library provided in ApPy is more powerful/less strict than the Autoprotocol spec, i.e. it has deserialization support for short-hand such as `ml`.
Ultimately, this still gets serialized into `milliliter` and not `ml`, so I think that's fine to provide that convenience to the writer. 
It'll become a concern if we do end up serializing to an unsupported unit.

Reply:
Broadly though, we've talked about potentially expanding the units in the spec to a bigger range. That's a separate topic which has some lively discussion, though we haven't seen real world needs for that yet.

Reply:
As a specification I am implementing according to what is listed at http://autoprotocol.org/specification/. Here there is no mention of `ml`. We must call it out. Let's down the line if I am implementing this specification in say Java, I might implement to support only `meter, millimeter, micrometer`.